apollo-federation/src/query_graph/graph_path/transition.rs

```diff
@@ -26,3 +55,508 @@ impl GraphPathTriggerVariant for QueryGraphEdgeTransition {
+                    // interface to another subgraph (the tail is not the interface object anymore),
+                    // we need to "restore" the type condition first.
+                    let updated_path = self.advance_with_transition(
+                        &QueryGraphEdgeTransition::Downcast {
+                            source: source.clone(),
+                            from_type_position: tail_type.clone(),
+                            to_type_position: field_definition_position.parent().clone(),
+                        },
+                        condition_resolver,
+                        override_conditions,
+                    )?;
+                    // The case we described above should be the only case we capture here, and so
+                    // the current subgraph must have the implementation type (it may not have the
+                    // field we want, but it must have the type) and so we should be able to advance
+                    // to it.
+                    let Either::Left(updated_path) = updated_path else {
+                        bail!(
+                            "Advancing {} for {} unexpectedly gave unadvanceables",
+                            self,
+                            transition,
+                        );
+                    };
+                    // Also note that there is currently no case where we should have more than one
+                    // option.
+                    let num_options = updated_path.len();
+                    let Some(updated_path) = iter_into_single_item(updated_path.into_iter()) else {
+                        bail!(
+                            "Advancing {} for {} unexpectedly gave {} options",
+                            self,
+                            transition,
+                            num_options,
+                        );
+                    };
+                    to_advance = Cow::Owned(updated_path);
+                    // We can now continue on dealing with the actual field.
+                }
+            }
+        }
+
+        let mut options: Vec<TransitionGraphPath> = vec![];
+        let mut dead_end_closures: Vec<UnadvanceableClosure> = vec![];
+
+        for edge in to_advance.next_edges()? {
+            let edge_weight = to_advance.graph.edge_weight(edge)?;
+            // The edge must match the transition. If it doesn't, we cannot use it.
+            if !edge_weight
+                .transition
+                .matches_supergraph_transition(transition)?
+            {
+                continue;
+            }
+
+            if let Some(override_condition) = &edge_weight.override_condition {
+                if !edge_weight.satisfies_override_conditions(override_conditions) {
+                    let graph = to_advance.graph.clone();
+                    let override_condition = override_condition.clone();
+                    let override_conditions = override_conditions.clone();
```

Comment:
Looks like we will need to wrap `override_conditions` with `Rc` (or `Arc`) when we optimize the code.
Also, we can capture the `edge` here, instead of `override_conditions`. But, these can be done later when we optimize this code. I think it's worth creating a separate Jira ticket for optimization.

Reply:
Good callout, it was simple enough that I just did it now (the closure re-reads the `edge`'s `override_condition`, and `override_conditions` is in an `Arc`).